# Guided project 1:
This porject is about to help the developer estimate how much revenue of each of our apps in App Store and Google Play Store based on how many customers of each app and how often they see and engage with the app, which also help to understand which type of apps are likely to attract more users on Google Play and the App Store.

- We'll start by opening and exploring these two data sets. To make them easier to explore, we created a function named explore_data() that you can repeatedly use to print rows in a readable way.

In [1]:
def explore_data(dataset, start, end, row_and_columns = False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') # add a new (emty) line after each row
        
    if row_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))        

Takes in four parameters:
    - dataset, which will be a list of lists
    - start and end, which will both be integers and represent the starting and the ending indices of a slice from the dataset
    - rows_and_columns, which will be a Boolean and has False as a default argument

Slices the dataset using dataset[start:end]

Loops through the slice, and for each iteration, prints a row and adds a new line after that row using print('\n')
    - The \n in print('\n') is a special character that won't print. Instead, the \n character adds a new line, and we use print('\n') to add some blank space between rows

Prints the number of rows and columns if rows_and_columns is True
    - dataset shouldn't have a header row, or the function will print the wrong number of rows (one more row compared to the actual length)

To help you better understand what print('\n') does, we printed three rows from the AppleStore.csv dataset below. In the first code cell, we don't use print('\n') between rows, while in the second one we do:


In [2]:
opened_file_apple = open('F:\DATA\Data scientist\From DATAQUEST\data practice\AppleStore.csv', encoding = 'utf8')
opened_file_google = open('F:\DATA\Data scientist\From DATAQUEST\data practice\googleplaystore.csv', encoding = 'utf8')
from csv import reader
read_file_apple = reader(opened_file_apple)
read_file_google = reader(opened_file_google)
apps_data_apple = list(read_file_apple)
apps_data_google = list(read_file_google)

In [3]:
explore_data(apps_data_apple, 1,3)
len(apps_data_apple[1:])
print('***')
explore_data(apps_data_google, 1,3)
len(apps_data_google[1:])


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


***
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']




10841

In [4]:
print('Column name of AppleStore data:')
print(apps_data_apple[0])
print('\n')
print('Column name of Google data')
print(apps_data_google[0])

Column name of AppleStore data:
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Column name of Google data
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


## Check for and delete a row with incorrect data

In [5]:
for row in apps_data_google:
    if len(row) != len(apps_data_google[0]):
        print(row)
        print('\n')
        print('Index position is:', apps_data_google[1:].index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Index position is: 10472


In [6]:
for row in apps_data_google:
    header_length = len(apps_data_google[0])
    row_length = len(row)
    if row_length != header_length:
        print(row)
        print(apps_data_google.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10473


In [7]:
print(apps_data_google[1:][10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [8]:
del apps_data_google[10473]

In [9]:
print(apps_data_google[10473])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


## Remove the duplicates

In [10]:
unique_apps = []
duplicate_apps = []

for app in apps_data_google:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('Number of unique apps: ', len(unique_apps))
print('\n')
print('Number of duplicates: ', len(duplicate_apps))
print('\n')
print('Example of duplicate apps:')
print(duplicate_apps[3:6])

Number of unique apps:  9660


Number of duplicates:  1181


Example of duplicate apps:
['ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box']


We don't want to count certain apps more than once when we analyze data, so we need to remove duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, however, when we examine the rows we printed for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to number of reviews. The difference numbers show the data collected at different times.

We can use this information to build a criteration for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given apps.

In [11]:
print('Expected length: ', len(apps_data_google[1:]) - 1181)

Expected length:  9659


In [12]:
reviews_max = {}

for row in apps_data_google[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max and n_reviews > reviews_max[name]:
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews

len(reviews_max)
print(reviews_max['Instagram'])
print('Check with result: 66577446')

66577446.0
Check with result: 66577446


In [13]:
android_clean = [] # store our new cleaned data set
already_added = [] # store app names

for row in apps_data_google[1:]:
    name = row[0]
    n_reviews = float(row[3])
#    print(n_reviews)
#    print(name)
#    print(type(reviews_max[name]))
    if n_reviews == reviews_max[name] and (name not in already_added): # Supplementary condition : 'name not in already_added'
        android_clean.append(row)
        already_added.append(row[0]) # this help us to keep track of apps that we already added

print(len(android_clean))

9659


We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). if we just check for 'reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [14]:
# An example of the app that has multiple entries with the number of reviews the same as the highest number of reviews
for row in apps_data_google:
    if row[0] == 'Box':
        print(row)

['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [15]:
print(apps_data_apple[:3])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']]


## Remove apps that are not designed for English- speaking audience

In [16]:
print(apps_data_apple[1:][813][1])
print(apps_data_apple[1:][6731][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ


* A function to check whether these app names are detected as English or non-English

In [17]:
def check_non_English(a_string):
    for character in a_string:
        if ord(character) > 127:
            return print('Not in English')
    
    print('English app')

In [18]:
check_non_English('Instagram')

English app


In [19]:
check_non_English('爱奇艺PPS -《欢乐颂2》电视剧热播')

Not in English


In [20]:
check_non_English('Docs To Go™ Free Office Suite')

Not in English


In [21]:
check_non_English('Instachat 😜')

Not in English


In [22]:
def is_English(a_string):
    non_ascii = 0
    for character in a_string:
        if ord(character) > 127:
            non_ascii += 1
    if non_ascii > 3:
        return False
    else:
        return True

In [23]:
is_English('Docs To Go™ Free Office Suite')

True

In [24]:
is_English('Instachat 😜')

True

In [25]:
is_English('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [26]:
android_clean_English = [] # store our new cleaned data set
ios_clean_English = [] # store our new cleaned data set

for row in android_clean:
    name = row[0]
    if is_English(name):
        android_clean_English.append(row)

for row in apps_data_apple[1:]:
    name = row[1]
    if is_English(name):
        ios_clean_English.append(row)
        
print('ANDROID')
explore_data(android_clean_English, 0, 3, True)
print('\n')
print('IOS')
explore_data(ios_clean_English, 0, 3, True)


ANDROID
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


IOS
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24

# Isolating free apps

In [27]:
print(apps_data_apple[0])
print('\n')
print(apps_data_google[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


### one way to isolate free apps

In [28]:
free_English_android = []
free_English_ios = []

for row in ios_clean_English:
    price_split = row[4].split('$')
    if len(price_split) > 2:
        price = float(price_split[1])
    else:
        price = float(price_split[0])
#    print(price)
    if price == 0:
        free_English_ios.append(row)

for row in android_clean_English:
    price_split = row[7].split('$')
    if len(price_split) > 1:
        price = float(price_split[1])
    else:
        price = float(price_split[0])
    if price == 0:
        free_English_android.append(row)
        
explore_data(free_English_ios, 0, 3, True)
print('\n')
explore_data(free_English_android, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 3222
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Vari

### Simpler way to isolate free apps

In [29]:
android_final = []
ios_final = []

for app in android_clean_English:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_clean_English:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print(len(android_final))
print(len(ios_final))

8864
3222


# Most common apps by genre
As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps

To minimize out risks and overhead, our validation strategy for an app is comprised of three steps:
    1. Build a minimal Android version of the app, and add it to Google Play.
    2. If app has a good response from users, we then develop it further.
    3. If the app is profitable after six months, we also build a ios version of the app and add it to the App Store.
    
Because our end goal is to add the appon both the App Store and Google Play, we need to find app profiles that are successfully on both markets. For instance , a profile that might work well for both markets might be a productivity app that make use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Catergory columns of the Google Play data set.

## Frequency table
we'll build two functions we can use to analyze the frequency table:

    - One function to generate frequency tables that show percentages
    - Another function that we can use to display the percentages in a descending order

In [30]:
# Generate frequency tables that show percentages
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentage = {}
    for key in table:
        percentage = (table[key]/ total) * 100
        table_percentage[key] = percentage
    return table_percentage


In [31]:
# A function to display the percenatge in a descending order

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key],key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
                               

In [32]:
freq_ios_apps = {}
freq_android_apps = {}

for app in ios_final:
    name = app[1]
    if name in freq_ios_apps:
        freq_ios_apps[name] += 1
    else:
        freq_ios_apps[name] = 1

for app in android_final:
    name = app[0]
    if name in freq_android_apps:
        freq_android_apps[name] += 1
    else:
        freq_android_apps[name] = 1
        
print(len(freq_android_apps))
print(len(freq_ios_apps))

8864
3220


In [33]:
# As a part of our additional research, it turns out there are 2 app names with the frequency of more than onein Apple Store.
# Since they have different ids, we assume that they are different apps eventhough they have the same names.

for app in freq_ios_apps:
    if freq_ios_apps[app] > 1:
        print(app)
    

Mannequin Challenge
VR Roller Coaster


In [34]:
print(apps_data_apple[0][11])

prime_genre


In [35]:
display_table(ios_final, 11) # ios - prime_genres

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see that among the free English apps, more than half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of these apps are designed for education, followed by social networking apps which amount for 3.29% of apps in our dataset.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.) while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that the fun apps are the most numerous doesn't also imply that they also have the greatest number of users - the demand might not be the same as the offer.

In [36]:
display_table(android_final, 1) # android - category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tool, business, lifestyle, productivity, etc.). However, if we investigate it further , we can see that the family category (which account for almost 19% of the apps) means mostly games for kids.

In [37]:
print(apps_data_google[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [38]:
display_table(android_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category column is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture, so we're only working with the Category column moving forward.

Up on this point, we found that App Store is dominated by apps designed for fun, while Google Play shows a more balanced of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

# Most popuplar apps

## On Apple Store

One way to find out what genres are the most popular (have the most users) is to calculate the average of installs for each app genre. For Google Play data set, we can find this information in the 'Installs' column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the 'rating_count_tot' app.

In [39]:
print(apps_data_apple[0])
print('\n')
print(apps_data_google[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [40]:
print(apps_data_apple[0][-5], apps_data_apple[0][5])
print(apps_data_google[0][-4], apps_data_google[1][5])
print(apps_data_google[0][1], apps_data_google[0][5])

prime_genre rating_count_tot
Genres 10,000+
Category Installs


In [41]:
prime_genre_ios = freq_table(ios_final, -5)
prime_genre_google = freq_table(android_final, 1)

def average(unique_genres, dataset, index1, index2):
    average_list ={}
    for genre in unique_genres:
        total = 0
        len_genre = 0
        for row in dataset:
            genre_app = row[index1]
            if genre_app == genre:
                n_ratings_float = float(row[index2])
                total += n_ratings_float
                len_genre += 1
        average = total/len_genre 
        a_tuple = (genre, average)
        average_list[genre] = average
    return average_list

In [42]:
def sort_tuple(tuple):
    a_list = []    
    for key in tuple:
        a_tuple = (tuple[key],key)
        a_list.append(a_tuple)
        
    a_new_list = sorted(a_list, reverse = True)
    for a_value in a_new_list:
        print(a_value[1],':',a_value[0])

In [43]:
sorted_ios = average(prime_genre_ios, ios_final,-5, 5)
sort_tuple(sorted_ios)

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


On average, navigation apps have the highest number of user reviews, this figure is heavily influenced by Waze and Google Maps, which have close to half million user reviews together

In [44]:
for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1],':',app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social netwroking apps, where average is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the average, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary which skew up the average rating:

In [45]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1],':',app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


However , this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from books, an audio version of the book, etc. On top of that, we could alse embed a dictionary with the app, so user don't need to exit out app to look for word in an external app.

This ideas seems to fit well with the fact that the App Store is dominated by for_fund apps. This suggests the market might be a bit saturated with for_fun apps, which means a pratical app might have more of a chance to stand out among the huge number of apps on Apple Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seems to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

    - Weather apps - people generally don't spend too much time in-app, the chances of making profit from in-apps adds are low. Also, getting a reliable live weather data may require us to connect our apps to non-free APIs.
   
    - Food and drink - examples here inlcude Starbucks, Duckin' Donuts, McDonald's etc. So making a popular food and drink apps requires actual cooking and a delivery service, which is outside the scope of our company.
    
    - Finance apps - these apps involve banking, paying bills, money transfer, etc. Building a finance app require domain knowledge, and we don't want to find a finance expert just to build an app.
 

## On Google Play

In [46]:
display_table(android_final, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


Eventhough we have data about the number of installs for Google Play market, so we should be able to get a clearer picture about the genre popularity. However, the install numbers don't seem precise enough - we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.)

For instance, we don't know whether an app with 100,000+ installs has 100.000 installs, 200,000 or 350,000. However, we don't need very precise data for our purposes - we only want to find out which app genres attract the most users.

We're going to leave the number as they are, which mean that we'll consider that app with 100,000+ installs has 100,000 installs, and an app with an 1,000,000+ installs has 1,000,000 installs, and so on. 

In [47]:
# one example of the data we will analyze and also to identify the correct index number of the columns.
print(android_final[1][1])
print(android_final[1][5])

ART_AND_DESIGN
5,000,000+


In [48]:
unique_genre_android = freq_table(android_final, 1)
average_data = {}

for category in unique_genre_android:
    total = 0
    len_category = 0
    for row in android_final:
        if row[1] == category:
            n_installs = row[5].replace(',','')
            n_installs = n_installs.replace('+','')
            n_installs = float(n_installs)
            total += n_installs
            len_category += 1
    average = total/len_category
    average_data[category] = average
    

In [49]:
sort_tuple(average_data)

COMMUNICATION : 38456119.167247385
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15588015.603248259
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5074486.197183099
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4056941.7741935486
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 51315

On average, Communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts) and a few others with over 100 and 500 million installs

In [50]:
for app in android_final:
     if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                       or app[5] == '500,000,000+'
                                       or app[5] == '100,000,000+'):
            print(app[0],':',app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would ve reduced roughly ten times

In [51]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',','')
    n_installs = n_installs.replace('+','')
    if app[1] == 'COMMUNICATION' and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m)/len(under_100_m)

3603485.3884615386

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Googe Photos and other popular photos editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.)

Again, the main concern is that these apps might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete agaist.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genres and their number of installs:

In [52]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0],':',app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [53]:
for app in android_final:
     if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                       or app[5] == '500,000,000+'
                                       or app[5] == '100,000,000+'):
            print(app[0],':',app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few popular apps, so this market still show potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle of popularity (between 1,000,000 and 100,000,000 downloads):

In [54]:
for app in android_final:
     if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                       or app[5] == '5,000,000+'
                                       or app[5] == '10,000,000+'
                                       or app[5] == '100,000,000'):
            print(app[0],':',app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which sugggest that building an app around a popular book can be profitable. It seems that taking a popular books (perhaps a more recent book) and turning it into an app could be profitable for both Google Play and App Store markets.

However, it looks like the market is already full of liraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the boook, a forum where people can dicuss the book, etc.

# Conclusion
In this project, we analyzed data about the App Store and Google Play mobile app with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and App Store market. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.